context.go

```diff
@@ -6,53 +6,142 @@ import (
 	"flag"
 	"fmt"
 	"strings"
+	"time"
 )
 
-// Context is a type that is passed through to
-// each Handler action in a cli application. Context
+// Context is an interface that is passed through to
+// each Handler action in a cli application. It
 // can be used to retrieve context-specific args and
 // parsed command-line options.
-type Context struct {
-	context.Context
-	App           *App
-	Command       *Command
+type Context interface {
+	// Functions to be able to manage Context type
+	App() *App
+	SetCommand(command *Command)
+	Command() *Command
+	Context() context.Context
+	ParentContext() Context
+	SetShellComplete(shellComplete bool)
+	ShellComplete() bool
+	SetFlagSet(set *flag.FlagSet)
+	FlagSet() *flag.FlagSet
+	NumFlags() int
+	Set(name, value string) error
+	IsSet(name string) bool
+	LocalFlagNames() []string
+	FlagNames() []string
+	Lineage() []Context
+	Value(name string) interface{}
+	Args() Args
+	NArg() int
```

Comment:
This is a great example of something we can probably remove to shrink our interface.

No reason to keep an `NArg()` function when we could just do `c.Args().Len()`

Reply:
I agree. However, I think we must first begin with deprecating it in our first v3 release and then remove it in v3.1 which will allow people to upgrade without having to face issues.

![Screenshot 2020-01-29 at 6 19 59 PM](https://user-images.githubusercontent.com/25657923/73358137-fc77ff80-42c3-11ea-840f-a66176bf3d98.png)

The deprecation notice is quiet clear in most IDEs.

Reply:
It's more important to adhere to semantic versioning here. Removing a deprecated feature is still a breaking change, meaning people will no longer safely be able to go from `3.0` to `3.1`.

If we're interested in marking it as deprecated, we should do so in the `v2` branch now, or mark it in `v3` and remove it in `v4`. That said, this kind of change is trivial enough we'd probably be fine just bundling it with the rest.

Reply:
Hmm. What we can do is we deprecate things during development time and then remove deprecated things before making the final release.